src/random.h

```diff
@@ -13,33 +13,81 @@
 #include <stdint.h>
 #include <limits>
 
-/* Seed OpenSSL PRNG with additional entropy data */
-void RandAddSeed();
+/**
+ * Overall design of the RNG and entropy sources.
+ *
+ * We maintain a single global 256-bit RNG state for all high-quality randomness.
+ * The following (classes of) functions interact with that state by mixing in new
+ * entropy, and optionally extracting random output from it:
+ *
+ * - The GetRand*() class of functions, as well as construction of FastRandomContext objects,
+ *   perform 'fast' seeding, consisting of mixing in:
+ *   - A stack pointer (indirectly committing to calling thread and call stack)
+ *   - A high-precision timestamp (rdtsc when available, c++ high_resolution_clock otherwise)
+ *   - Hardware RNG (rdrand) when available.
+ *   These entropy sources are very fast, and mostly designed to protect against situations
+ *   where a VM state restore/copy results in multiple systems with the same randomness.
```

Comment:
> Does this refer to an accidental scenario, where a user ends up generating the same private keys or accidentally signing with the same nonce? Or also a malicious scenario?

Does it matter whether it can be triggered accidentally or intentionally?

> So do I understand correctly that the reason FastRandomContext needs to fast is (mostly) this, or are there also areas of code that need the additional performance?

I don't understand the question or the relation with the code you're commenting on.

Reply:
I was trying to understand _why_ it needs to be fast. The explanation here suggests it's mostly about protecting against some sort of VM replay scenario. But in later commits it's clear that there's also a performance concern, e.g. the Perfmon can take two seconds.

Reply:
Ah, I see. I think you're looking at it the wrong way. Of course there are many places in the code where we need fast randomness (of various sizes); the issue that the speed concern means we can't do very good seeding, so we need a separate function for the few places where we _really_ need good randomness (GetStrongRandBytes).

Reply:
I do understand that fast also implies weak. The question is why you would ever accept weak, if you already code from strong. The obvious answer is performance. But the comment here talks about a VM replay scenario.

But now I see what confused me: "These entropy sources are very fast, and mostly designed to protect against situations  where a VM state restore" - the part after "and" doesn't refer to "fast"; they are an _additional_ benefit of these sources.

Reply:
FastButStillHardToReplayRandomContext :-P

Reply:
No, `FastRandomContext` does *not* protect against replay. `GetRand*()` does.